tests/bluetooth/bsim_bt/bsim_test_notify/src/gatt_server_test.c

```diff
@@ -0,0 +1,210 @@
+				 sizeof(long_chrc_data));
+}
+
+static void short_subscribe(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	bool notif_enabled = (value == BT_GATT_CCC_NOTIFY);
+
+	printk("Short notifications %s\n", notif_enabled ? "enabled" : "disabled");
+}
+
+static void long_subscribe(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	bool notif_enabled = (value == BT_GATT_CCC_NOTIFY);
+
+	printk("Long notifications %s\n", notif_enabled ? "enabled" : "disabled");
+}
+
+BT_GATT_SERVICE_DEFINE(test_svc, BT_GATT_PRIMARY_SERVICE(TEST_SERVICE_UUID),
+		       BT_GATT_CHARACTERISTIC(TEST_CHRC_UUID,
+					      BT_GATT_CHRC_NOTIFY | BT_GATT_CHRC_READ,
+					      BT_GATT_PERM_READ, read_test_chrc, NULL, NULL),
+		       BT_GATT_CCC(short_subscribe, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
+		       BT_GATT_CHARACTERISTIC(TEST_LONG_CHRC_UUID,
+					      BT_GATT_CHRC_NOTIFY | BT_GATT_CHRC_READ,
+					      BT_GATT_PERM_READ, read_long_test_chrc, NULL, NULL),
+		       BT_GATT_CCC(long_subscribe, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE));
+
+struct notification_meta {
+	size_t len;
+};
+
+static volatile size_t num_notifications_sent;
+
+static void notification_sent(struct bt_conn *conn, void *user_data)
+{
+	struct notification_meta *meta = user_data;
+
+	printk("Sent notification with length %d\n", meta->len);
+	num_notifications_sent++;
+}
+
+static inline void short_notify(void)
+{
+	static struct notification_meta meta = {
+		.len = CHRC_SIZE,
+	};
+	static struct bt_gatt_notify_params params = {
+		.attr = &attr_test_svc[1],
+		.data = chrc_data,
+		.len = CHRC_SIZE,
+		.func = notification_sent,
+		.user_data = &meta,
+		.uuid = NULL,
+	};
+	int err = bt_gatt_notify_cb(g_conn, &params);
+
+	if (err < 0) {
+		FAIL("Short notify failed (err %d)\n", err);
+	}
+}
+
+static inline void long_notify(void)
+{
+	static struct notification_meta meta = {
+		.len = LONG_CHRC_SIZE,
+	};
+	static struct bt_gatt_notify_params params = {
+		.attr = &attr_test_svc[4],
+		.data = long_chrc_data,
+		.len = LONG_CHRC_SIZE,
+		.func = notification_sent,
+		.user_data = &meta,
+		.uuid = NULL,
+	};
+	int err = bt_gatt_notify_cb(g_conn, &params);
+
+	if (err < 0) {
+		FAIL("Long notify failed (err %d)\n", err);
+	}
+}
+
+static void test_main(void)
+{
+	int err;
+	const struct bt_data ad[] = {
+		BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	};
+
+	err = bt_enable(NULL);
+	if (err != 0) {
+		FAIL("Bluetooth init failed (err %d)\n", err);
+		return;
+	}
+
+	printk("Bluetooth initialized\n");
+
+	err = bt_le_adv_start(BT_LE_ADV_CONN_NAME, ad, ARRAY_SIZE(ad), NULL, 0);
+	if (err != 0) {
+		FAIL("Advertising failed to start (err %d)\n", err);
+		return;
+	}
+
+	printk("Advertising successfully started\n");
+
+	WAIT_FOR_FLAG(flag_is_connected);
+
+	while (bt_eatt_count(g_conn) < CONFIG_BT_EATT_MAX) {
+		k_sleep(K_MSEC(100));
+	}
+	printk("EATT connected\n");
+
+	k_sleep(K_SECONDS(1));
+
+	for (int i = 0; i < NOTIFICATIONS/2; i++) {
```

Comment:
Do we need to ensure that `NOTIFICATIONS` is an even number? 
```suggestion
	for (int i = 0; i < NOTIFICATIONS / 2; i++) {
```

Reply:
Yes. I may have to rewrite this if we change to `K_NO_WAIT` instead of `K_FOREVER` as pointed out by Alexander further up and sending this may fail.